src/realm/sync/network/default_socket.cpp

```diff
@@ -65,25 +65,90 @@ class DefaultWebSocketImpl final : public WebSocketInterface, public Config {
     void websocket_read_error_handler(std::error_code ec) override
     {
         m_logger.error("Reading failed: %1", ec.message()); // Throws
-        m_observer.websocket_read_or_write_error_handler(ec);
+        constexpr bool was_clean = false;
+        websocket_error_and_close_handler(was_clean, Status{ErrorCodes::ReadError, ec.message()});
     }
     void websocket_write_error_handler(std::error_code ec) override
     {
         m_logger.error("Writing failed: %1", ec.message()); // Throws
-        m_observer.websocket_read_or_write_error_handler(ec);
+        constexpr bool was_clean = false;
+        websocket_error_and_close_handler(was_clean, Status{ErrorCodes::WriteError, ec.message()});
     }
     void websocket_handshake_error_handler(std::error_code ec, const HTTPHeaders*,
                                            const std::string_view* body) override
     {
-        m_observer.websocket_handshake_error_handler(ec, body);
+        ErrorCodes::Error error;
+        bool was_clean;
+
+        if (ec == websocket::Error::bad_response_3xx_redirection ||
+            ec == websocket::Error::bad_response_301_moved_permanently ||
+            ec == websocket::Error::bad_response_308_permanent_redirect) {
+            error = ErrorCodes::WebSocket_MovedPermanently;
+            was_clean = true;
+        }
+        else if (ec == websocket::Error::bad_response_401_unauthorized) {
+            error = ErrorCodes::WebSocket_Unauthorized;
+            was_clean = true;
+        }
+        else if (ec == websocket::Error::bad_response_403_forbidden) {
+            error = ErrorCodes::WebSocket_Forbidden;
+            was_clean = true;
+        }
+        else if (ec == websocket::Error::bad_response_5xx_server_error ||
+                 ec == websocket::Error::bad_response_500_internal_server_error ||
+                 ec == websocket::Error::bad_response_502_bad_gateway ||
+                 ec == websocket::Error::bad_response_503_service_unavailable ||
+                 ec == websocket::Error::bad_response_504_gateway_timeout) {
+            error = ErrorCodes::WebSocket_InternalServerError;
+            was_clean = false;
+        }
+        else {
+            error = ErrorCodes::WebSocket_Protocol_Mismatch;
```

Comment:
Use Abnormal Closure here with was clean = false and if we get past line 111: `if (i != std::string_view::npos) {`, then set either the protocol mismatch, too old or too new with was_clean = true.

Reply:
Here's an example of where this is reporting the wrong error:
```
[2023/01/20 22:20:49.597] 3: Connection[1]: Closing the websocket with status='WebSocket: Protocol Mismatch: Bad WebSocket response 404 not found', was_clean='true'
[2023/01/20 22:20:49.597] 3: Connection[1]: Connection closed due to error
[2023/01/20 22:20:49.597] 3: An unexpected sync error was caught during test: 'app destroyed during token refresh': 'Protocol version negotiation failed: No version supported by both client and server' for '/data/mci/5807aadb9a4c7148b7137b6611a20667/tmp/realm_KeQHYr/realm.dGY61r'
```
from: https://spruce.mongodb.com/task/realm_core_stable_ubuntu2004_ubsan_object_store_tests_patch_ead23bd8258d1a4e76aef2cc66a914d23cc18efd_63cb0c2b3066151c43994a69_23_01_20_21_48_27/tests?execution=0&sortBy=STATUS&sortDir=ASC

Reply:
Ah, yes. I didn't really know what to use as default to make it fatal as it was before. If we use Abnormal Closure, we'll retry it (non-fatal)..

Reply:
I'll use the new `Fatal` here, since Abnormal Closure would mean a retry and that should not be the case here.